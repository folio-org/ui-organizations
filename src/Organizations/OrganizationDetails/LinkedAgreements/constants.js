import { FormattedMessage } from 'react-intl';

export const AGREEMENTS_LIST_COLUMNS = {
  name: 'name',
  startDate: 'startDate',
  endDate: 'endDate',
  agreementStatus: 'agreementStatus',
};

export const AGREEMENTS_LIST_VISIBLE_COLUMNS = [
  AGREEMENTS_LIST_COLUMNS.name,
  AGREEMENTS_LIST_COLUMNS.startDate,
  AGREEMENTS_LIST_COLUMNS.endDate,
  AGREEMENTS_LIST_COLUMNS.agreementStatus,
];

export const AGREEMENTS_LIST_COLUMNS_MAPPING = {
  [AGREEMENTS_LIST_COLUMNS.name]: <FormattedMessage id="ui-organizations.linkedAgreements.agreement.name" />,
  [AGREEMENTS_LIST_COLUMNS.startDate]: <FormattedMessage id="ui-organizations.linkedAgreements.agreement.startDate" />,
  [AGREEMENTS_LIST_COLUMNS.endDate]: <FormattedMessage id="ui-organizations.linkedAgreements.agreement.endDate" />,
  [AGREEMENTS_LIST_COLUMNS.agreementStatus]: <FormattedMessage id="ui-organizations.linkedAgreements.agreement.agreementStatus" />,
};
