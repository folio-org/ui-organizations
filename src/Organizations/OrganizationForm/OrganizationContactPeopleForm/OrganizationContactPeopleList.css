.ellipsis {
  width: 12em;
}
