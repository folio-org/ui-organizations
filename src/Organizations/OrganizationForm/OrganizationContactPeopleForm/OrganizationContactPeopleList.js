import React from 'react';
import PropTypes from 'prop-types';
import {
  get,
  map,
  sortBy,
} from 'lodash';
import {
  FormattedMessage,
  useIntl,
} from 'react-intl';

import {
  Button,
  Icon,
  MultiColumnList,
} from '@folio/stripes/components';
import { Pluggable } from '@folio/stripes/core';
import { acqRowFormatter } from '@folio/stripes-acq-components';

import { transformCategoryIdsToLables } from '../../../common/utils/category';

const columnMapping = {
  contactCategories: <FormattedMessage id="ui-organizations.contactPeople.categories" />,
  contactEmails: <FormattedMessage id="ui-organizations.contactPeople.emails" />,
  contactName: <FormattedMessage id="ui-organizations.contactPeople.name" />,
  unassignContact: null,
};

const visibleColumns = ['contactName', 'contactCategories', 'contactEmails', 'unassignContact'];

const getContactsUrl = (orgId, contactId) => {
  const ending = contactId ? `/contacts/${contactId}/view` : '/contacts/add-contact';
  const starting = orgId ? `/organizations/${orgId}` : '/organizations';

  return `${starting}${ending}`;
};

const AddContactButton = ({ fetchContacts, fields, stripes, orgId }) => {
  const addContacts = (contacts = []) => {
    const addedContactIds = new Set(fields.getAll());
    const newContactsIds = map(contacts.filter(({ id }) => !addedContactIds.has(id)), 'id');

    if (newContactsIds.length) {
      fetchContacts([...addedContactIds, ...newContactsIds]);
      newContactsIds.forEach(contactId => fields.push(contactId));
    }
  };

  const renderNewContactBtn = () => {
    const url = `/organizations${orgId ? '/' + orgId : ''}/contacts/add-contact`;

    return (
      <Button
        marginBottom0
        buttonStyle="primary"
        to={url}
      >
        <FormattedMessage id="ui-organizations.contactPeople.newContact" />
      </Button>
    );
  };

  return (
    <Pluggable
      aria-haspopup="true"
      type="find-contact"
      dataKey="contact"
      searchLabel={<FormattedMessage id="ui-organizations.contactPeople.addContact" />}
      searchButtonStyle="default"
      disableRecordCreation
      stripes={stripes}
      addContacts={addContacts}
      renderNewContactBtn={renderNewContactBtn}
    >
      <span data-test-add-contact>
        <FormattedMessage id="ui-organizations.contactPeople.noFindContactPlugin" />
      </span>
    </Pluggable>
  );
};

AddContactButton.propTypes = {
  fetchContacts: PropTypes.func.isRequired,
  fields: PropTypes.object,
  stripes: PropTypes.object,
  orgId: PropTypes.string,
};

const alignRowProps = { alignLastColToEnd: true };

const OrganizationContactPeopleList = ({ fetchContacts, fields, contactsMap, orgId, categoriesDict, stripes }) => {
  const intl = useIntl();
  const contentData = sortBy(fields.getAll().map((contactId, _index) => ({
    ...get(contactsMap, contactId, {}),
    _index,
  })), [({ firstName }) => firstName.toLowerCase()]);

  const anchoredRowFormatter = ({ rowProps, ...rest }) => {
    return acqRowFormatter({
      ...rest,
      rowProps: {
        ...rowProps,
        to: getContactsUrl(orgId, rest.rowData.id),
      },
    });
  };

  const resultsFormatter = {
    contactCategories: ({ categories = [] }) => transformCategoryIdsToLables(categoriesDict, categories),
    contactEmails: ({ emails }) => map(emails, 'value').join(', '),
    contactName: contact => `${contact.firstName} ${contact.lastName}`,
    unassignContact: (contact) => (
      <Button
        align="end"
        aria-label={intl.formatMessage({ id: 'ui-organizations.contacts.button.unassign' })}
        buttonStyle="fieldControl"
        data-test-unassign-contact
        type="button"
        onClick={(e) => {
          e.preventDefault();
          fields.remove(contact._index);
        }}
      >
        <Icon icon="times-circle" />
      </Button>
    ),
  };

  return (
    <>
      <MultiColumnList
        id="contact-list"
        columnMapping={columnMapping}
        contentData={contentData}
        formatter={resultsFormatter}
        rowFormatter={anchoredRowFormatter}
        rowProps={alignRowProps}
        visibleColumns={visibleColumns}
      />
      <br />
      <AddContactButton
        fetchContacts={fetchContacts}
        fields={fields}
        stripes={stripes}
        orgId={orgId}
      />
    </>
  );
};

OrganizationContactPeopleList.propTypes = {
  fetchContacts: PropTypes.func.isRequired,
  fields: PropTypes.object,
  orgId: PropTypes.string,
  categoriesDict: PropTypes.arrayOf(PropTypes.object),
  contactsMap: PropTypes.object,
  stripes: PropTypes.object,
};

export default OrganizationContactPeopleList;
