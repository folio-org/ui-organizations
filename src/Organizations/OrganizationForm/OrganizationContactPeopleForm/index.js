export { default as OrganizationContactPeopleForm } from './OrganizationContactPeopleForm';
