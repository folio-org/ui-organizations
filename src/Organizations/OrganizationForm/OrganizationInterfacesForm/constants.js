export const columnMapping = {
  name: 'Name',
  url: 'URL',
  unassign: null,
};

export const visibleColumns = ['name', 'url', 'unassign'];
