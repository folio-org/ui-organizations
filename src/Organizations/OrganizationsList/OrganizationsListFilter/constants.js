import React from 'react';
import { FormattedMessage } from 'react-intl';

import {
  ORGANIZATION_STATUS,
  ORGANIZATION_STATUS_LABELS,
} from '../../../common/constants';

export const FILTERS = {
  STATUS: 'status',
  TAGS: 'tags',
  ADDRESS_COUNTRY: 'address_country',
  LANGUAGE: 'language',
  PAYMENT_METHOD: 'paymentMethod',
  STATS_AVAILABLE: 'statsAvailable',
  IS_VENDOR: 'isVendor',
};

export const STATUS_OPTIONS = Object.values(ORGANIZATION_STATUS).map(status => ({
  value: status,
  label: ORGANIZATION_STATUS_LABELS[status],
}));

export const BOOLEAN_OPTIONS = [
  {
    value: 'true',
    label: <FormattedMessage id="ui-organizations.filterConfig.boolean.true" />,
  },
  {
    value: 'false',
    label: <FormattedMessage id="ui-organizations.filterConfig.boolean.false" />,
  },
];
