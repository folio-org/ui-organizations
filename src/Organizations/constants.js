import React from 'react';
import { FormattedMessage } from 'react-intl';

export const ORGANIZATION_SECTIONS = {
  summarySection: 'summarySection',
  contactInformationSection: 'contactInformationSection',
  contactPeopleSection: 'contactPeopleSection',
  interfacesSection: 'interfacesSection',
  vendorInformationSection: 'vendorInformationSection',
  vendorTermsSection: 'agreementsSection',
  accountsSection: 'accountsSection',
  notesSection: 'notesSection',
  integrationDetailsSection: 'integrationDetailsSection',
};

export const ORGANIZATION_SECTION_LABELS = {
  [ORGANIZATION_SECTIONS.summarySection]: <FormattedMessage id="ui-organizations.summary" />,
  [ORGANIZATION_SECTIONS.contactInformationSection]: <FormattedMessage id="ui-organizations.contactInformation" />,
  [ORGANIZATION_SECTIONS.contactPeopleSection]: <FormattedMessage id="ui-organizations.contactPeople" />,
  [ORGANIZATION_SECTIONS.interfacesSection]: <FormattedMessage id="ui-organizations.interface" />,
  [ORGANIZATION_SECTIONS.vendorInformationSection]: <FormattedMessage id="ui-organizations.vendorInformation" />,
  [ORGANIZATION_SECTIONS.vendorTermsSection]: <FormattedMessage id="ui-organizations.vendorTerms" />,
  [ORGANIZATION_SECTIONS.accountsSection]: <FormattedMessage id="ui-organizations.accounts" />,
  [ORGANIZATION_SECTIONS.integrationDetailsSection]: <FormattedMessage id="ui-organizations.integrationDetails" />,
};

export const EDI_CODE_TYPES = [
  { label: '31B (US-SAN)', value: '31B/US-SAN' },
  { label: '014 (EAN)', value: '014/EAN' },
  { label: '091 (Supplier-assigned ID)', value: '091/Vendor-assigned' },
  { label: '092 (Library-assigned ID)', value: '092/Customer-assigned' },
];

export const FTP_TYPES = [
  { label: 'SFTP', value: 'SFTP' },
  { label: 'FTP', value: 'FTP' },
];

export const TRANSMISSION_MODES = [
  { label: 'ASCII', value: 'ASCII' },
  { label: 'Binary', value: 'Binary' },
];

export const CONNECTION_MODES = [
  { label: 'Active', value: 'Active' },
  { label: 'Passive', value: 'Passive' },
];

export const CREATE_UNITS_PERM = 'organizations.acquisitions-units-assignments.assign';
export const MANAGE_UNITS_PERM = 'organizations.acquisitions-units-assignments.manage';

// Mapping between attribute (field) in form and id of accordion
export const MAP_FIELD_ACCORDION = {
  name: ORGANIZATION_SECTIONS.summarySection,
  code: ORGANIZATION_SECTIONS.summarySection,
  status: ORGANIZATION_SECTIONS.summarySection,
  aliases: ORGANIZATION_SECTIONS.summarySection,
  addresses: ORGANIZATION_SECTIONS.contactInformationSection,
  phoneNumbers: ORGANIZATION_SECTIONS.contactInformationSection,
  urls: ORGANIZATION_SECTIONS.contactInformationSection,
  agreements: ORGANIZATION_SECTIONS.vendorTermsSection,
  accounts: ORGANIZATION_SECTIONS.accountsSection,
};
