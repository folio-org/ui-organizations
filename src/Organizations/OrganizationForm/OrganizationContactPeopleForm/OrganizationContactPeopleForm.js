import React, { useCallback, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { FieldArray } from 'redux-form';

import {
  Col,
  Loading,
  Row,
} from '@folio/stripes/components';
import {
  stripesConnect,
  stripesShape,
} from '@folio/stripes/core';
import { batchFetch } from '@folio/stripes-acq-components';

import {
  baseContactsResource,
  categoriesResource,
} from '../../../common/resources';
import { DICT_CATEGORIES } from '../../../common/constants';
import OrganizationContactPeopleList from './OrganizationContactPeopleList';

function OrganizationContactPeopleForm({ open, orgId, mutator, storedContactIds, stripes }) {
  const [categoriesDict, setCategoriesDict] = useState();
  const [contacts, setContacts] = useState();
  const [isLoading, setIsLoading] = useState(true);

  const fetchContacts = useCallback(ids => {
    setIsLoading(true);
    batchFetch(mutator.contactsManualFetch, ids)
      .then(setContacts)
      .catch(() => {
        setContacts([]);
      })
      .finally(() => setIsLoading(false));
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    mutator[DICT_CATEGORIES].GET().then(setCategoriesDict);
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => fetchContacts(storedContactIds), [fetchContacts, storedContactIds]);

  if (isLoading || !categoriesDict || !open) {
    return (
      <Loading />
    );
  }

  const contactsMap = contacts.reduce((acc, contact) => {
    acc[contact.id] = contact;

    return acc;
  }, {});

  return (
    <Row>
      <Col xs={12}>
        <FieldArray
          name="contacts"
          component={OrganizationContactPeopleList}
          props={{
            fetchContacts,
            categoriesDict,
            contactsMap,
            orgId,
            stripes,
          }}
        />
      </Col>
    </Row>
  );
}

OrganizationContactPeopleForm.manifest = Object.freeze({
  [DICT_CATEGORIES]: {
    ...categoriesResource,
    accumulate: true,
    fetch: false,
  },
  contactsManualFetch: {
    ...baseContactsResource,
    accumulate: true,
    fetch: false,
  },
});

OrganizationContactPeopleForm.propTypes = {
  mutator: PropTypes.object.isRequired,
  open: PropTypes.bool,
  orgId: PropTypes.string,
  storedContactIds: PropTypes.arrayOf(PropTypes.string),
  stripes: stripesShape.isRequired,
};

export default stripesConnect(OrganizationContactPeopleForm);
