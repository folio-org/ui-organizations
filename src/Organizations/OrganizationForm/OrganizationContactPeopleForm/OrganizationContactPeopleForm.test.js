import React from 'react';
import { render, screen } from '@folio/jest-config-stripes/testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import stripesFinalForm from '@folio/stripes/final-form';

import { contact } from 'fixtures';

import { DICT_CATEGORIES } from '../../../common/constants';
import OrganizationContactPeopleForm from './OrganizationContactPeopleForm';

const mutatorMock = {
  contactsManualFetch: {
    GET: jest.fn().mockReturnValue(Promise.resolve([contact])),
  },
  [DICT_CATEGORIES]: {
    GET: jest.fn().mockReturnValue(Promise.resolve([])),
  },
};
const TestForm = stripesFinalForm({})(
  () => {
    return (
      <form>
        <OrganizationContactPeopleForm
          open
          stripes={{}}
          storedContactIds={[contact.id]}
          mutator={mutatorMock}
        />
      </form>
    );
  },
);

const defaultProps = {
  initialValues: {
    contacts: [contact.id],
  },
};
const renderForm = (props = defaultProps) => render(
  <TestForm
    onSubmit={jest.fn()}
    {...props}
  />,
  { wrapper: MemoryRouter },
);

describe('OrganizationContactPeopleForm', () => {
  beforeEach(() => {
    global.document.createRange = global.document.originalCreateRange;
  });

  afterEach(() => {
    global.document.createRange = global.document.mockCreateRange;
  });

  it('should render correct structure', async () => {
    const { asFragment } = renderForm();

    await screen.findByText('ui-organizations.contactPeople.name');

    expect(asFragment()).toMatchSnapshot();
  });
});
