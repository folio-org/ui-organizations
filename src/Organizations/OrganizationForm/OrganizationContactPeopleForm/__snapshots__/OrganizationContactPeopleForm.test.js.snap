// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrganizationContactPeopleForm should render correct structure 1`] = `
<DocumentFragment>
  <form>
    <div
      class="row"
    >
      <div
        class="col-xs-12"
      >
        <div
          aria-live="assertive"
          aria-relevant="additions"
          class="sr-only"
          tabindex="-1"
        />
        <div
          aria-rowcount="2"
          class="mclContainer"
          data-total-count="0"
          id="contact-list"
          role="grid"
          style="position: relative; overflow: hidden; height: auto; width: 100%;"
          tabindex="0"
        >
          <div
            class="mclHeaderContainer"
          >
            <div
              aria-rowindex="1"
              class="mclHeaderRow"
              role="row"
              style="display: flex; width: 17px;"
            >
              <div
                aria-sort="none"
                class="mclHeader"
                id="list-column-contactname"
                role="columnheader"
                style="flex: 0 0 auto;"
              >
                ui-organizations.contactPeople.name
              </div>
              <div
                aria-sort="none"
                class="mclHeader"
                id="list-column-contactcategories"
                role="columnheader"
                style="flex: 0 0 auto;"
              >
                ui-organizations.contactPeople.categories
              </div>
              <div
                aria-sort="none"
                class="mclHeader"
                id="list-column-contactemails"
                role="columnheader"
                style="flex: 0 0 auto;"
              >
                ui-organizations.contactPeople.emails
              </div>
              <div
                aria-sort="none"
                class="mclHeader"
                id="list-column-unassigncontact"
                role="columnheader"
                style="flex: 0 0 auto;"
              />
            </div>
          </div>
          <div
            class="mclScrollable"
            style="position: static; width: 100%;"
          >
            <div
              class="mclRowContainer"
              role="rowgroup"
              style="position: static;"
            >
              <div
                aria-rowindex="2"
                class="mclRowFormatterContainer"
                data-row-index="row-0"
                role="row"
                style="position: static;"
              >
                <a
                  aria-label=""
                  class="mclRow mclIsOdd mclIsInteractive alignLastColToEnd"
                  data-row-inner="0"
                  href="/organizations/contacts/e0fb5df2-cdf1-11e8-a8d5-f2801f1b9fd1/view"
                  style="min-width: 0px;"
                  tabindex="0"
                >
                  <div
                    class="mclCell mclCellStyle"
                    role="gridcell"
                  >
                    Brownstein, Carrie
                  </div>
                  <div
                    class="mclCell mclCellStyle"
                    role="gridcell"
                  />
                  <div
                    class="mclCell mclCellStyle"
                    role="gridcell"
                  >
                    cust.service03@amazon.com
                  </div>
                  <div
                    class="mclCell mclCellStyle"
                    role="gridcell"
                  >
                    <button
                      aria-label="ui-organizations.contacts.button.unassign"
                      class="button fieldControl align-end"
                      data-test-unassign-contact="true"
                      type="button"
                    >
                      <span
                        class="inner"
                      >
                        <span>
                          Icon
                        </span>
                      </span>
                    </button>
                  </div>
                </a>
              </div>
              <div
                class="mclEndOfListContainer"
                data-end-of-list="1"
              >
                <div
                  class="mclCenteredContainer"
                  style="width: 100%; visibility: hidden; height: 0px; padding: 0px;"
                >
                  <div
                    class="textCentered"
                    data-test-layout="true"
                  >
                    <div
                      class="mclEndOfList"
                      data-test-layout="true"
                    >
                      <span>
                        Icon
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <br />
        <span
          data-test-add-contact="true"
        >
          ui-organizations.contactPeople.noFindContactPlugin
        </span>
      </div>
    </div>
  </form>
  <div />
</DocumentFragment>
`;
